Implement access to `device` and `dtype`

SCiarella opened the pull request:
Closes issue #23

This PR exposes `device` and `dtype` such that they can be controlled and set externally by the user. This is particularly relevant for running `diffwofost` on GPU by setting `device="cuda"`. There are example of this in the notebooks and in the documentation.

The tests have been update by adding a new `@pytest.fixture` that checks if a GPU is available, and if so, it re-runs the most important tests on the GPU a second time. Notice that GitHub actions do not provide GPU access (for free), so I have been using my local hardware to test the code.

-----

Notice that without finalizing the Engine, the GPU device could be slower than the CPU, due to temporary fixes like the use of `deepcopy` in the `TestEngine`.

fnattino commented on docs/notebooks/optimization.ipynb:
The notebook fails because of the changes introduced in #59 (the engine now allows in-memory configuration, so the .conf files have been removed). The fix will be included as part of #69, so we can drop the changes to this notebook.

fnattino commented on pyproject.toml:
```diff
@@ -61,6 +61,9 @@ Issues = "https://github.com/WUR-AI/diffwofost/issues"
 
 [tool.pytest.ini_options]
 testpaths = ["tests"]
+filterwarnings = [
```
Oh, so clean this can be configured here!

fnattino commented on tests/physical_models/crop/test_leaf_dynamics.py:
```diff
@@ -130,7 +139,10 @@ def test_leaf_dynamics_with_one_parameter_vector(self, param):
+        ) = prepare_engine_input(
+            test_data, crop_model_params, meteo_range_checks=False, device=device
+        )
+        # config_path = str(phy_data_folder / "WOFOST_Leaf_Dynamics.conf")
```
```suggestion
```

fnattino commented on tests/physical_models/crop/test_phenology.py:
```diff
@@ -74,23 +78,28 @@ def __init__(
         # pass new value of parameters to the model
         for name, value in params_dict.items():
+            if isinstance(value, torch.Tensor) and value.device.type != self.device:
+                value = value.to(self.device)
```
Same as for `test_leaf_dynamics`:
```suggestion
```

fnattino commented on tests/physical_models/crop/test_root_dynamics.py:
```diff
@@ -420,23 +438,31 @@ class TestDiffRootDynamicsGradients:
-        assert loss.grad_fn is None
+        # If there is no graph at all, then there is no gradient by definition.
+        if not loss.requires_grad:
+            return
```
Uhm, this conditional essentially passes without any assertion. In order to make this test more effective, should one split the parameters that we know give an empty graph from the parameters that give a zero gradient, and assert that the right situation is encountered?

fnattino commented on src/diffwofost/physical_models/crop/phenology.py:
```diff
@@ -417,9 +533,46 @@ def initialize(self, day, kiosk, parvalues):
+        Vernalisation.device = self.device
+        Vernalisation.dtype = self.dtype
+
+        # Helpler function to cast and broadcast all parameters to params_shape.
```
Does this function need to be a closure or can it be written as a regular method?
 
Spotted a typo:
```suggestion
        # Helper function to cast and broadcast all parameters to params_shape.
```

SCiarella commented on tests/physical_models/crop/test_leaf_dynamics.py:
```diff
@@ -46,17 +44,21 @@ def __init__(
         # pass new value of parameters to the model
         for name, value in params_dict.items():
+            if isinstance(value, torch.Tensor) and value.device.type != self.device:
+                value = value.to(self.device)
```
not needed with centralized control

SCiarella commented:
Thanks @fnattino for the careful review!

I have implemented most of the suggestions that you proposed. In particular, I introduced a new class in `config.py` called `ComputeConfig` that is used to set globally `dtype` and `device`. Using this global control, I was able to remove most of the redundant type checks that you also pointed out. @SarahAlidoost what do you think of this solution?

There is a point that remains open regarding the `kiosk`. Currently, I have implemented some type/device conversion for the variables that come from the kiosk. Later, we should ensure that the engine enforces global types and devices, such that the kiosk does not need to be converted every time.
